x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/AllFirstAllLastTestingUtils.java

```diff
@@ -0,0 +1,160 @@
+        for (Page page : input) {
+            Block values = page.getBlock(0);
+            LongBlock timestamps = page.getBlock(1);
+
+            for (int p = 0; p < page.getPositionCount(); ++p) {
+                Tuple<List<Long>, List<Object>> pair = unpack(timestamps, values, p);
+                work.add(pair.v1(), pair.v2());
+            }
+        }
+    }
+
+    /**
+     * Unpacks the timestamps and values, of the passed blocks at the given position, into two lists.
+     */
+    private static Tuple<List<Long>, List<Object>> unpack(LongBlock timestamps, Block values, int p) {
+        // extract all timestamps at this position into a list
+        List<Long> timestampsAtPosition = BlockTestUtils.valuesAtPosition(timestamps, p, true);
+
+        // extract all values at this position into a list
+        List<Object> valuesAtPosition = BlockTestUtils.valuesAtPosition(values, p, true);
+
+        return Tuple.tuple(timestampsAtPosition, valuesAtPosition);
+    }
+
+    /**
+     * This class servers as a manual aggregator for ALL_FIRST/ALL_LAST aggregator functions, that is used as the ground truth
+     * during unit tests.
+     */
+    static class GroundTruthFirstLastAggregator {
+        /**
+         * A container for all values associated with the best timestamp, across all pages and positions. Elements can be either single
+         * values (i.e.: {@code Object}), or multivalues (i.e.: {@code List<Object>}). In the latter case, the list must be sorted to
+         * allow for proper comparisons later during the check phase, which also performs the same operation if the result from the real
+         * aggregator was multivalued.
+         */
+        private final Set<Object> expectedValues = new HashSet<>();
+
+        /**
+         * Whether this is a FIRST or LAST aggregator
+         */
+        private final boolean first;
+
+        /**
+         * The best timestamp
+         */
+        private Long expectedTimestamp;
+
+        /**
+         * Whether an observation has been made
+         */
+        private boolean observed;
+
+        GroundTruthFirstLastAggregator(boolean first) {
+            this.first = first;
+            this.observed = false;
+        }
+
+        /**
+         * This method is called exactly once per position, as all timestamps and values for that position are passed in here.
+         */
+        void add(List<Long> timestamps, List<Object> values) {
+            if (timestamps.isEmpty()) {
+                // no-op: a null timestamp sorts last so it can't beat any existing timestamp
+                return;
+            }
+            long timestamp = first ? Collections.min(timestamps) : Collections.max(timestamps);
+            if (observed == false || (first ? timestamp < expectedTimestamp : timestamp > expectedTimestamp)) {
+                expectedTimestamp = timestamp;
+                expectedValues.clear();
+                values.sort(null);
```

Comment:
Above, you say that values can be an object, but it looks like you'll get a singleton list instead. That's *fine*, but I think the comment is wrong.